src/staff_rename.lua

```diff
@@ -206,30 +255,111 @@ function staff_rename()
+        end
+        add_ctrl(dialog, "horizontalline", "", 0, row[2] + 8, 0, col_w * 3.5 + 20, 0, 0)
+        str.LuaString = "*Custom*"
+        master_autonumber_popup:AddString(str)
+
         --local h_line = add_ctrl(dialog, "horizontalline", 0, row[1], 1, col_w * 3, 0, 0)
         --
         for i, j in pairs(staves) do
-            static_staff[i] = add_ctrl(dialog, "static", staves[i], 10, row[i + 1], row_h, col_w, 0, 0)
-            edit_fullname[i] = add_ctrl(dialog, "edit", fullnames[i], col[1], row[i + 1], row_h, col_w, 0, 0)
-            edit_abbname[i] = add_ctrl(dialog, "edit", abbnames[i], col[2], row[i + 1], row_h, col_w, 0, 0)
-            copy_button[i] = add_ctrl(dialog, "button", "→", col[2] - col_gap + 2, row[i + 1], row_h-4, 16, 0, 0)
+            static_staff[i] = add_ctrl(dialog, "static", staves[i], 10, row[i + 2], row_h, col_w, 0, 0)
+            edit_fullname[i] = add_ctrl(dialog, "edit", fullnames[i], col[1], row[i + 2], row_h, col_w, 0, 0)
+            edit_abbname[i] = add_ctrl(dialog, "edit", abbnames[i], col[2], row[i + 2], row_h, col_w, 0, 0)
+            copy_button[i] = add_ctrl(dialog, "button", "→", col[2] - col_gap + 2, row[i + 2], row_h-4, 16, 0, 0)
+            autonumber_check[i] = add_ctrl(dialog, "checkbox", "", col[3] + autox_width, row[i+2], row_h, 13, 0, 0)
+            autonumber_popup[i] = add_ctrl(dialog, "popup", "", col[3] + 60, row[i+2], row_h, col_w - 20, 0, 0)
+            for key, val in pairs(autonumber_style_list) do
+                str.LuaString = autonumber_style_list[key]
+                autonumber_popup[i]:AddString(str)
+            end
+            if autonumber_bool[i] then
+                autonumber_check[i]:SetCheck(1)
+                autonumber_popup[i]:SetEnable(true)
+            else
+                autonumber_check[i]:SetCheck(0)
+                autonumber_popup[i]:SetEnable(false)
+                master_autonumber_check:SetCheck(0)
+                master_autonumber_popup:SetEnable(false)
+            end
+            autonumber_popup[i]:SetSelectedItem(autonumber_style[i])
             row_count = row_count + 1
         end
         --
-        local form_select = add_ctrl(dialog, "popup", "", col[1], row[row_count + 1] + row_h/2, row_h, col_w - col_gap, 0, 0)
-        local forms = {"Inst. in Tr.","Tr. Inst."}
+        local form_select = add_ctrl(dialog, "popup", "", col[1], row[row_count + 2] + row_h/2, row_h, col_w - col_gap, 0, 0)
+        local forms = {"Instrument in Trn.","Trn. Instrument"}
         for i,j in pairs(forms) do
             str.LuaString = forms[i]
             form_select:AddString(str)
-        end
+        end   
+        local hardcode_autonumber_btn = add_ctrl(dialog, "button", "Hardcode Autonumbers", col[3] + autox_width, row[row_count + 2], row_h, col_w, 0, 0)
         --
         dialog:CreateOkButton()
         dialog:CreateCancelButton()
         --
+        function hardcode_autonumbers()

```

Comment:
Not worth changing right now, but a general rule of thumb is to not define functions inside other functions (unless it's a short, simple, one-off function). It works, but it's harder to maintain.

The reason is that functions defined inside other functions are harder to reason about how they work. Ideally, we want a function to be as predictable as possible, and that often means writing "pure" functions.

A pure function is one that meets the following criteria:

1) It's behavior depends solely on it's arguments
2) It does not have side effects<sup>*</sup>

Here's a more rigorous definition: https://en.wikipedia.org/wiki/Pure_function

By defining a function inside another function, that inner function has access to all variables that exist in that outer function and can therefore mutate any of the outer function's variables. This can lead to incredibly unpredictable behavior whenever someone wants to update the script. So it's best not to mutate the outer function's variables.

The best to prevent that is to define functions at the top-level, so there is no nesting. Then you can't mutate an outer function's variables!

Likewise, it's best not to use the outer function's variables in the first place. Because then you essentially have implicit arguments to the function. It looks like I should be able to just call `hardcode_autonumbers()` to run this function. But if it depends on any variables in the outer function, then changing those variables outside the function will then change the behavior of `hardcode_autonumbers()`. That's very unpredictable! So it's best not to even use them.

And if you aren't accessing or mutating an outer function's variables, there's no reason to define a function inside another function.

<sup>*</sup> In Finale Lua code, it's difficult to not have any side effects since we are constantly updating FCNotes, FCEntries, and other Finale objects. In this repository, I think that's perfectly acceptable and actually a good practice to allow those mutations. But mutating variables outside the function's scope is a bad practice.